Capture caller frame of napari.run() when starting the console.

Carreau opened the pull request:
Should fix napari/napari#4098
Replaces napari/napari#4140, and see discussion there as well
Needs napari/napari#4212

alisterburt reviewed (approved):
napari/napari#4212 has merged now, should we merge this too @Carreau? Changes look good to me, although there are a few strange extra blank lines :)

psobolewskiPhD commented:
I installed this PR branch and napari `main`. No more errors, but I must be missing something, because, I made a script:
```
import napari
a = 3
b = 4
viewer = napari.Viewer()
c = 5
napari.run()
```
I was expecting `c` to be available in the napari console, but it's not.
🤔

Carreau commented:
as we inspect frames my guess is we see the frame at the point the viewer
was created so before C ?

On Wed, Dec 21, 2022 at 21:48 Peter Sobolewski ***@***.***>
wrote:

> I installed this PR branch and napari main. No more errors, but I must be
> missing something, because, I made a script:
>
> import napari
>
> a = 3
>
> b = 4
>
> viewer = napari.Viewer()
>
> c = 5
>
> napari.run()
>
>
> I was expecting c to be available in the napari console, but it's not.
> 🤔
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/napari/napari-console/pull/18#issuecomment-1362075315>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACR5TZJI3BOYHNNT6OGJLLWONULLANCNFSM5QDP6SXA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

psobolewskiPhD commented:
Ah, I assumed it would be stuff after the viewer.
But checking again, `a` and `b` arn't available either.
☹️

Carreau commented:
Yeah, I forgot to push...

Carreau commented:
Or technically I pushed, but did not commit, so it pushed nothing.

brisvag commented:
> I was expecting `c` to be available in the napari console, but it's not.

that would be great too, but probably more complicated cause it requires back and forth communication?

psobolewskiPhD commented:
> Or technically I pushed, but did not commit, so it pushed nothing.

Ah! I'm so pleased it wasn't me screwing something up again! 😊 
Anyhow, refreshing the branch and it works perfectly:
```
a
Out[1]: 3

b
Out[2]: 4

c
Out[3]: 5
```

Czaki commented on napari_console/qt_console.py:
```diff
@@ -55,6 +58,14 @@ def str_to_rgb(arg):
 
+def _not_napari(n: int, frame: FrameType):
+    # in-n-out is used in napari for dependency injection.
+    for pref in ["napari", "in_n_out"]:
```
What with modules named `napari-something`?

jni replied:
should we change these to 
```suggestion
    for pref in ["napari.", "in_n_out."]:
```
then?

Czaki reviewed (commented):
I have mixed feelings. It is really nice in the context of scripts or notebooks. 

But for example, PartSeg allow opening napari viewer from the view menu. And with this PR the context of this method is pushed: 

https://github.com/4DNucleome/PartSeg/blob/3f8f3bc56e1fa055f65d428b1d9e6d776af8ae37/package/PartSeg/common_gui/main_window.py#L230-L235

I do not feel that it is proper behavior.

psobolewskiPhD commented:
@czaki can you elaborate what is the issue with your PartSeg use case?
Does it maybe make sense to have the console disabled in that case?

Czaki commented:
> @Czaki can you elaborate what is the issue with your PartSeg use case?

exporting internal application variables to the console. (It could be easily workaround - for a person like me who understands napari internals)

> Does it maybe make sense to have the console disabled in that case?

I use it regularly in the context of debugging and development. And some person uses it to perform actions not possible by GUI (ex layer scale).

psobolewskiPhD commented:
Gotcha. Makes sense. Maybe we need an argument to *not* do this when launching napari?
Even from a script someone may not want the variables passed?

jni commented:
Can we try to push this over the line? Todos:

- check my [suggestion on the prefixes](https://github.com/napari/napari-console/pull/18#discussion_r1070150857)
- figure out solution for libraries that use napari as an internal tool; two half-baked suggestions: 
  - add a "max_depth=1" kwarg for the frame, so that this only works for scripts (depth=0) and a single function (depth=1), but is easily configurable if we find that doesn't cover everything. (Not sure whether this works in IPython?)
  - add a "blacklist" of modules where we don't do this, including partseg

For point 2, I wouldn't mind leaving it to a later PR, as this is already extremely useful.

Czaki commented:
I like option 1. 

I'm not the favorite of option 2. It should be rather information in the documentation that will inform about this and suggest ways to workaround it. (maybe such variable `_not_incude_in_napari_console=True`).

Carreau commented:
I updated this PR. 

`max_depth` , I changed to min_depth, I'm guessing what you want is to be able to define `def my_lancher(...)`, and capture the context of my_launcher  ? In general frames depth here are going to be weird, as there is a lot of internal napari frames.

And in general I would suggest people embedding this to just subclass and overwrite the frame predicate to know when to stop. Like setting a predcate that just always reuturn `True # in napari`,  will thus not embed any namespace.

I added the `.` at the end of module names.

jni reviewed (approved):
❤️‍🔥 Awesome, thank you so much for the update @Carreau!

Czaki reviewed (dismissed):
I have installed this PR. Launch the console and print `locals()` 

Ia has found the following variables: `args`, `kwargs,` `cls`, `possibly_static_exec`, `TYPE_CHECKING`, and a few more. 

So something is wrong, but until Friday, I will not have time to investigate this PR.

jni commented:
How are you running the viewer? When I run it with `examples/add_labels.py`, I get the following (in addition to some private ones):

```
 'data': <module 'skimage.data' from '/Users/jni/projects/scikit-image/skimage/data/__init__.py'>,
 'threshold_otsu': <function skimage.filters.thresholding.threshold_otsu(image=None, nbins=256, *, hist=None)>,
 'label': <function skimage.measure._label.label(label_image, background=None, return_num=False, connectivity=None)>,
 'closing': <function skimage.morphology.gray.closing(image, footprint=None, out=None)>,
 'remove_small_objects': <function skimage.morphology.misc.remove_small_objects(ar, min_size=64, connectivity=1, *, out=None)>,
 'square': <function skimage.morphology.footprints.square(width, dtype=<class 'numpy.uint8'>, *, decomposition=None)>,
 'clear_border': <function skimage.segmentation._clear_border.clear_border(labels, buffer_size=0, bgval=0, mask=None, *, out=None)>,
 'napari': <module 'napari' from '/Users/jni/projects/napari/napari/__init__.py'>,
 'image': array([[158, 141, 160, ..., 142, 199, 182],
        [162, 153, 168, ..., 193, 203, 136],
        [178, 173, 173, ..., 143, 179, 120],
        ...,
        [ 98, 117, 143, ...,  64,  65,  66],
        [128, 125, 153, ...,  63,  64,  63],
        [ 99, 102, 132, ...,  66,  62,  59]], dtype=uint8),
 'thresh': 112,
 'bw': array([[ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        ...,
        [ True,  True,  True, ..., False, False, False],
        [ True,  True,  True, ..., False, False, False],
        [False, False,  True, ..., False, False, False]]),
 'cleared': array([[False, False, False, ..., False, False, False],
        [False, False, False, ..., False, False, False],
        [False, False, False, ..., False, False, False],
        ...,
        [False, False, False, ..., False, False, False],
        [False, False, False, ..., False, False, False],
        [False, False, False, ..., False, False, False]]),
 'label_image': array([[0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0],
        [0, 0, 0, ..., 0, 0, 0]], dtype=int32),
 'viewer': Viewer(camera=Camera(center=(0.0, 101.0, 141.5), zoom=2.0461631546384247, angles=(0.0, 0.0, 90.0), perspective=0.0, mouse_pan=True, mouse_zoom=True), cursor=Cursor(position=(185.75141043603875, 0.3028212509458994), scaled=True, style=<CursorStyle.STANDARD: 'standard'>, size=10.0), dims=Dims(ndim=2, ndisplay=2, order=(0, 1), axis_labels=('0', '1'), range=(RangeTuple(start=0.0, stop=202.0, step=1.0), RangeTuple(start=0.0, stop=283.0, step=1.0)), margin_left=(0.0, 0.0), margin_right=(0.0, 0.0), point=(101.0, 141.0), last_used=0), grid=GridCanvas(stride=1, shape=(-1, -1), enabled=False), layers=[<Image layer 'coins' at 0x2a53c0760>, <Labels layer 'segmentation' at 0x107558f70>], help='use <7> for transform, use <1> for activate the label eraser, use <2> for activate the paint brush, use <3> for activate the polygon tool, use <4> for activate the fill bucket, use <5> for pick mode', status='', tooltip=Tooltip(visible=False, text=''), theme='dark', title='napari', mouse_over_canvas=False, mouse_move_callbacks=[], mouse_drag_callbacks=[], mouse_double_click_callbacks=[], mouse_wheel_callbacks=[<function dims_scroll at 0x17196a200>], _persisted_mouse_event={}, _mouse_drag_gen={}, _mouse_wheel_gen={}, keymap={}),
 'label_layer': <Labels layer 'segmentation' at 0x107558f70>,
 'action_manager': <napari.utils.action_manager.ActionManager at 0x1695d6170>,
```

The only dodgy one is `action_manager` at the bottom, dunno what's going on there... 😅 But everything else is as expected. If you run it as main, then I wouldn't expect this PR to be doing much useful...

Czaki commented:
I run it with `napari`in shell

jni commented:
> I run it with `napari` in shell

I mean, in that situation I wouldn't expect this PR to be useful for anything, but I guess you're saying it should not in that case pollute the namespace? That probably requires a fix in our main function, or wherever we instantiate the console, rather than here?

Czaki commented:
> guess you're saying it should not in that case pollute the namespace

yes

Czaki commented:
Ok. I push commit with the proposition of a solution. 

There are two sources of problems. 
The first is caused by `CallerFrame` do not stop even on `__main__.py` (and then `run.py`). Ic could also be fixed on Napari itself, but we do not want this PR on the napari release. 

Also the name of `napari/__main__.py` is `__main__`, not `napari.__main__` 

I also allow disabling of collecting variables by creating the global variable `NAPARI_EMBED`. What did you think about this?

jni commented:
I'm totally fine with your proposal @Czaki

jni commented:
If you yourself are happy with your proposal can you dismiss your review and we can merge? 😃

jni commented:
Well the tests seem unhappy... 😅

Czaki commented:
> Well the tests seem unhappy... 😅

I forget that there is a global state in the console, so names in tests cannot colide.

jni commented:
I updated the README message to include a usage example.

Czaki commented on README.md:
```diff
@@ -20,6 +20,41 @@ and review the napari docs for plugin developers:
+`napari`, and `in_n_out` modules, and passing the variables in the frame's
+`f_locals` and `f_globals` to the console namespace.
+
+If you want to disable this behavior (for example, because you are embedding
+napari and the console within some larger application), you can add
+`NAPARI_EMBED=1` to your environment variables before instantiating the
+console.
```
This is not true. I implement it as global module-level variable, not an environment variable. Should I change it?

jni replied:
Oh! Sorry. I did read that and thought you misspoke. 😅 I think I prefer this way because it's easier to describe and quite a common way to change functionality. Otherwise you have to make sure that the global is accessible in the correct frame, which is not something people are used to thinking about, I think? 🤷

Czaki replied:
yes, but environment variable is global per process, and module level variable is in module. 

If someone would like first spawn PartSeg and then napari, If I use an environment variable, then it will prevent variable propagation in follow-up napari start. 

If using module-level variable it impacts only possible console spawn when open console from napari embedded here. But will not impact napari spawned from the top-level script.

brisvag commented on napari_console/qt_console.py:
```diff
@@ -140,6 +158,29 @@ def __init__(self, viewer: 'napari.viewer.Viewer'):
 
+    def _in_napari(self, n: int, frame: FrameType):
+        """
+        Predicates that return Wether we are in napari by looking
+        at:
+            1) the frames modules names:
+            2) the min_depth
```
```suggestion
        Determines whether we are in napari by looking at:
            1) the frames modules names:
            2) the min_depth
```

jni commented:
> The only dodgy one is action_manager at the bottom, dunno what's going on there...

(from https://github.com/napari/napari-console/pull/18#issuecomment-1758781001)

@Carreau it looks like that was added by you in napari/napari#2677 without any discussion. I see now in the PR description your comment, "I also inject action_manager in the qtconsole context, so one does not have to import it everytime", but in general I think the action manager is not really a user-facing object, even though it is convenient for people developing specific things. Was that for debugging and can it be removed? (This would be for 0.5.0 or later in any case...)

Carreau commented:
> I see now in the PR description your comment, "I also inject action_manager in the qtconsole context, so one does not have to import it everytime", 


I belive this was done before we had the shortcut editor so we needed a way to edit shortcuts is a relatively friendly way.

jni commented:
Well, I suggest that we merge this and fix the action_manager injection in another PR.

jni commented:
(It would be cool to simul-release this improvement with 0.4.19!)

jni commented:
Shit, I should have tested this again before merging 😅

Right now when I run `python `[`examples/add_labels.py`](https://github.com/napari/napari/blob/e8c41abf306e45dc06f8569f6bdd461386f15e92/examples/add_labels.py), I see the viewer, np, napari, and some IPython stuff:

```
In [1]: locals()
Out[1]: 
{'__name__': '__main__',
 '__doc__': 'Automatically created module for IPython interactive environment',
 '__package__': None,
 '__loader__': None,
 '__spec__': None,
 '__builtin__': <module 'builtins' (built-in)>,
 '__builtins__': <module 'builtins' (built-in)>,
 '_ih': ['', 'locals()'],
 '_oh': {},
 '_dh': [PosixPath('/Users/jni/projects/napari')],
 'In': ['', 'locals()'],
 'Out': {},
 'get_ipython': <bound method InteractiveShell.get_ipython of <ipykernel.inprocess.ipkernel.InProcessInteractiveShell object at 0x301fd1310>>,
 'exit': <IPython.core.autocall.ZMQExitAutocall at 0x3030068d0>,
 'quit': <IPython.core.autocall.ZMQExitAutocall at 0x3030068d0>,
 'open': <function io.open(file, mode='r', buffering=-1, encoding=None, errors=None, newline=None, closefd=True, opener=None)>,
 '_': '',
 '__': '',
 '___': '',
 'viewer': Viewer(camera=Camera(center=(0.0, np.float64(101.0), np.float64(141.5)), zoom=np.float64(1.7528169014084505), angles=(0.0, 0.0, 90.0), perspective=0.0, mouse_pan=True, mouse_zoom=True), cursor=Cursor(position=(1.0, 1.0), scaled=True, style=<CursorStyle.STANDARD: 'standard'>, size=np.float64(10.0)), dims=Dims(ndim=2, ndisplay=2, order=(0, 1), axis_labels=('0', '1'), rollable=(True, True), range=(RangeTuple(start=np.float64(0.0), stop=np.float64(202.0), step=np.float64(1.0)), RangeTuple(start=np.float64(0.0), stop=np.float64(283.0), step=np.float64(1.0))), margin_left=(0.0, 0.0), margin_right=(0.0, 0.0), point=(np.float64(101.0), np.float64(141.0)), last_used=0), grid=GridCanvas(stride=1, shape=(-1, -1), enabled=False), layers=[<Image layer 'coins' at 0x1750fa190>, <Labels layer 'segmentation' at 0x1750f9a50>], help='use <7> for transform, use <1> for activate the label eraser, use <2> for activate the paint brush, use <3> for activate the polygon tool, use <4> for activate the fill bucket, use <5> for pick mode', status='Ready', tooltip=Tooltip(visible=False, text=''), theme='dark', title='napari', mouse_over_canvas=False, mouse_move_callbacks=[], mouse_drag_callbacks=[], mouse_double_click_callbacks=[], mouse_wheel_callbacks=[<function dims_scroll at 0x1647e4040>], _persisted_mouse_event={}, _mouse_drag_gen={}, _mouse_wheel_gen={}, _keymap={}),
 'napari': <module 'napari' from '/Users/jni/projects/napari/napari/__init__.py'>,
 'action_manager': <napari.utils.action_manager.ActionManager at 0x15e051910>,
 'np': <module 'numpy' from '/Users/jni/micromamba/envs/all/lib/python3.11/site-packages/numpy/__init__.py'>,
 '_i': '',
 '_ii': '',
 '_iii': '',
 '_i1': 'locals()'}
```

I wonder if this commit broke things?

[5339095](https://github.com/napari/napari-console/pull/18/commits/53390956aba021a0bac9313078d6653deb37a2ba)

jni commented:
Indeed changing this:

```
if c.frame.f_globals.get("__name__", "") != "__main__" and "NAPARI_EMBED" not in c.frame.f_globals:
```

to

```
if "NAPARI_EMBED" not in c.frame.f_globals:
```

fixes it again. I'll make a PR, since I don't think the first part is necessary (?) @Czaki
